api/service/legacysync/epoch_syncing.go

```diff
@@ -0,0 +1,240 @@
+package legacysync
+
+import (
+	"sync"
+	"time"
+
+	"github.com/Workiva/go-datastructures/queue"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/harmony-one/harmony/consensus"
+	"github.com/harmony-one/harmony/core"
+	"github.com/harmony-one/harmony/core/types"
+	"github.com/harmony-one/harmony/internal/chain"
+	"github.com/harmony-one/harmony/internal/utils"
+	"github.com/harmony-one/harmony/node/worker"
+	"github.com/harmony-one/harmony/p2p"
+	"github.com/harmony-one/harmony/shard"
+	"github.com/pkg/errors"
+)
+
+type EpochSync struct {
+	beaconChain        *core.BlockChain
+	selfip             string
+	selfport           string
+	selfPeerHash       [20]byte // hash of ip and address combination
+	commonBlocks       map[int]*types.Block
+	lastMileBlocks     []*types.Block // last mile blocks to catch up with the consensus
+	syncConfig         *SyncConfig
+	isExplorer         bool
+	stateSyncTaskQueue *queue.Queue
+	syncMux            sync.Mutex
+	lastMileMux        sync.Mutex
+
+	syncStatus syncStatus
+}
+
+// GetSyncStatus get the last sync status for other modules (E.g. RPC, explorer).
+// If the last sync result is not expired, return the sync result immediately.
+// If the last result is expired, ask the remote DNS nodes for latest height and return the result.
+func (ss *EpochSync) GetSyncStatus() SyncCheckResult {
+	return ss.syncStatus.Get(func() SyncCheckResult {
+		return ss.isInSync(false)
+	})
+}
+
+// isInSync query the remote DNS node for the latest height to check what is the current
+// sync status
+func (ss *EpochSync) isInSync(doubleCheck bool) SyncCheckResult {
+	if ss.syncConfig == nil {
+		return SyncCheckResult{} // If syncConfig is not instantiated, return not in sync
+	}
+	otherHeight1 := getMaxPeerHeight(ss.syncConfig)
+	curBlock := ss.beaconChain.CurrentBlock()
```

Comment:
From the beaconChain, you only have epoch blocks, right? The EpochLastBlock of a epoch block is the epoch block itself.